src/v/kafka/server/handlers/fetch.cc

```diff
@@ -106,8 +125,22 @@ static ss::future<read_result> read_from_partition(
     std::unique_ptr<iobuf> data;
     std::vector<cluster::rm_stm::tx_range> aborted_transactions;
     try {
-        auto result = co_await rdr.reader.consume(
-          kafka_batch_serializer(), deadline ? *deadline : model::no_timeout);
+        const auto ts_deadline = deadline.value_or(model::no_timeout);
+        const auto failsafe_deadline = failsafe_deadline_from(ts_deadline);
+        auto consume_fut = rdr.reader.consume(
+          kafka_batch_serializer(), ts_deadline);
+        auto result
+          = co_await ss::with_timeout(failsafe_deadline, std::move(consume_fut))
+              .handle_exception_type(
+                [&reader_config](const ss::timed_out_error& err) {
```

Comment:
Do we think erroring out here will give us the information needed to diagnose the hang? I had sort of though the timeout logic would have been pushed down lower into the TS layer where we would spit out TS specific info which would help determine the cause for the hang. OTOH doing it here I guess guarantees to catch any kind of hang. Maybe the info we log about the reader config is enough?

Reply:
For tiered storage consumption involves creating the offset stream (which means downloading from cloud storage) and then the `remote_segment_batch_reader` working through the downloaded data, either of those may cause the time out but stream creation seems more likely. 

For stream creation we now log the client address. I will add the address to the batch reader logs as well, so if we get a timeout here in `fetch.cc`, we can correlate the address in the error with the logs from tiered storage (the stream creation and batch reader) and see where the code path was blocked when it timed out.

Reply:
added client address to the batch reader

Reply:
Sounds great, thanks!